src/services/development.ts

```diff
@@ -1,36 +1,102 @@
-import { TOKENS as $, services as prodServices } from './production'
-import { merge, build, ServiceDefinition } from './utils'
-import { mocks } from '@/api/mocks'
+import { setupWorker, RestHandler, MockedRequest } from 'msw'
+
+import { TOKENS as PROD_TOKENS, services as prodServices } from './production'
+import { merge, build, ServiceDefinition, token, get } from './utils'
+import { useBootstrap } from '../index'
+import { mocks, setupHandlers } from '@/api/mocks'
+import { fs, fakeApi } from '@/api/mocks/index'
 import CookiedEnv from '@/services/env/CookiedEnv'
-import KumaApi from '@/services/kuma-api/KumaApi'
-import { mockApi } from '@/services/kuma-api/MockKumaApi'
 import Logger from '@/services/logger/DatadogLogger'
 import { disabledLogger } from '@/services/logger/DisabledLogger'
+import type { TokenType } from '@/services/utils'
 
 export { constant, get, container, createInjections, build, merge } from './utils'
-export { TOKENS } from './production'
 
-export const services: ServiceDefinition[] = merge(prodServices, [
+type Msw = ReturnType<typeof setupWorker>
+const $ = {
+  ...PROD_TOKENS,
+  msw: token<Msw>('msw'),
+  mswKumaHandlers: token<RestHandler[]>('msw.kuma.handlers'),
+  mswKumaMocks: token<RestHandler[]>('msw.kuma.mocks'),
+  mswHandlers: token<RestHandler[]>('msw.handlers'),
+}
+
+type Env = TokenType<typeof $.Env>
+export const services: ServiceDefinition[] = [
+
   [$.Env, {
     service: CookiedEnv,
     arguments: [
       $.EnvVars,
     ],
   }],
-  [$.mocks, {
-    constant: mocks,
+
+  [$.logger, {
+    service: disabledLogger(Logger),
+    arguments: [
+      $.Env,
+    ],
   }],
-  [$.api, {
-    service: mockApi(KumaApi),
+
+  // Mock Service Worker
+  [$.msw, {
+    service: (handlers: RestHandler[]) => {
+      const worker = setupWorker(...handlers)
+      worker.start({
```

Comment:
**Issue** (maybe): This removes all the logic I added to ensure mocking works when providing msw with new handlers during runtime.

Specifically stopping the worker and starting it with new handlers was essential to make this work. With the removal of `MockKumaApi.ts`, I see no more logic that would ensure handlers are being set again when changing an API instance’s base URL.

Please especially test that this works.

Reply:
Have you tested that this works when changing the base URL during runtime?

Reply:
Do you mean the 'base' URL or the 'prefix' path?

If you mean can the prefix path change at runtime, then yes, tested. All works fine as far as I can see.

If you mean the base URL, then that doesn't really change at runtime as far as I understand it. It's set once, before MSW initialises and it stays the same throughout the lifetime of the application.

We need to clear up the difference between the base URL and a prefix path.

I know in the application we currently change the entire base URL over and over again, like there is no prefix path. But as this is the mocks/MSW, not the application, it works differently ie. `${baseURL}/${prefix}/${url}`

---

For the lifetime of MSW:

The baseURL doesn't change once set, ever.
The prefix path doesn't change, but contains a dynamic segment which is variable. Sometimes we use a prefix path, sometimes we don't (see next point for how we do this)
The url is our traditional URL, that doesn't change either but contains more dynamic segments that are variable. We sometimes dynamically prefix these URLs with the prefix path via a loop-de-loop through the majority of URLs from The Outside (i.e. from our service container).

---

If you think about it like that ^ it makes the whole thing an awful lot easier.

Hopefully that's clear enough for how I think about it, but let me know if you have any questions/queries or you need anything clarifying.

Reply:
All I wanted to know is if you tested changing API URLs to be reflected correctly in the mocks and you replied yes (be it in a very long manner).